internal/liveapply/liveapply.go

```diff
@@ -0,0 +1,119 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package liveapply
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/coreos/ignition/v2/config/util"
+	"github.com/coreos/ignition/v2/internal/exec"
+	"github.com/coreos/ignition/v2/internal/exec/stages"
+	_ "github.com/coreos/ignition/v2/internal/exec/stages/disks"
+	_ "github.com/coreos/ignition/v2/internal/exec/stages/fetch"
+	_ "github.com/coreos/ignition/v2/internal/exec/stages/fetch_offline"
+	_ "github.com/coreos/ignition/v2/internal/exec/stages/files"
+	_ "github.com/coreos/ignition/v2/internal/exec/stages/kargs"
+	_ "github.com/coreos/ignition/v2/internal/exec/stages/mount"
+	_ "github.com/coreos/ignition/v2/internal/exec/stages/umount"
+	"github.com/coreos/ignition/v2/internal/log"
+	"github.com/coreos/ignition/v2/internal/resource"
+	"github.com/coreos/ignition/v2/internal/state"
+
+	"github.com/coreos/ignition/v2/config/v3_4_experimental/types"
+)
+
+type Flags struct {
+	Root              string
+	IgnoreUnsupported bool
+	Offline           bool
+}
+
+func inContainer() bool {
+	if val, _ := os.LookupEnv("container"); val != "" {
+		return true
+	}
+
+	paths := []string{"/run/.containerenv", "/.dockerenv"}
+	for _, path := range paths {
+		if _, err := os.Stat(path); err == nil {
+			return true
+		}
+	}
+
+	return false
+}
+
+func Run(cfg types.Config, flags Flags, logger *log.Logger) error {
+	if !inContainer() {
+		return errors.New("this tool is not designed to run on a host system; reprovision the machine instead")
+	}
+
+	fetcher := resource.Fetcher{
+		Logger:  logger,
+		Offline: flags.Offline,
+	}
+
+	state := state.State{}
+	cfgFetcher := exec.ConfigFetcher{
+		Logger:  logger,
+		Fetcher: &fetcher,
+		State:   &state,
+	}
+
+	finalCfg, err := cfgFetcher.RenderConfig(cfg)
+	if err != nil {
+		return err
+	}
+
+	// verify upfront if we'll need networking but we're not allowed
+	if flags.Offline {
+		stage := stages.Get("fetch-offline").Create(logger, flags.Root, fetcher, &state)
+		if err := stage.Run(finalCfg); err != nil {
+			return err
+		}
+	}
+
+	// Order in which to apply live. This is overkill since effectively only
+	// `files` supports it right now, but let's be extensible. Also ensures that
+	// all stages are accounted for.
+	stagesOrder := []string{"fetch-offline", "fetch", "kargs", "disks", "mount", "files", "umount"}
+	allStages := stages.Names()
+	if len(stagesOrder) != len(allStages) {
+		panic(fmt.Sprintf("%v != %v", stagesOrder, allStages))
+	}
+
+	for _, stageName := range stagesOrder {
+		if !util.StrSliceContains(allStages, stageName) {
+			panic(fmt.Sprintf("stage '%s' invalid", stageName))
+		}
+		if strings.HasPrefix(stageName, "fetch") {
+			continue // already fetched
```

Comment:
We'd just call a no-op `ApplyLive`, right?  Maybe it'd be cleaner to drop this special case.